File: packages/localecheck-python/localecheck/vat.py
"""Tax-rate lookups: UK VAT by date, US sales tax by state.

This is the hand-curated, date-sensitive part — the bit a raw LLM cannot know
and the most defensible piece of the product.
"""
import datetime as _dt
import json
import os

_DATA = os.path.join(os.path.dirname(__file__), "data")


def _load(name):
    with open(os.path.join(_DATA, name), encoding="utf-8") as f:
        return json.load(f)


_UK = _load("uk_vat.json")
_US = _load("us_sales_tax.json")


def _uk_vat(date_str):
    try:
        d = _dt.date.fromisoformat(str(date_str))
    except ValueError as e:
        return {"ok": False, "reason": f"bad date: {e}"}
    history = _UK["standard_rate_history"]
    # find the last band whose start is <= date
    applicable = None
    for band in history:
        if _dt.date.fromisoformat(band["from"]) <= d:
            applicable = band
        else:
            break
    if applicable is None:
        return {"ok": False, "reason": f"no UK VAT rate recorded for {date_str} "
                f"(records start {history[0]['from']})"}
    return {"ok": True, "country": "GB", "tax_name": "VAT",
            "rate": applicable["rate"], "date": d.isoformat(),
            "effective_from": applicable["from"], "kind": "standard",
            "note": applicable.get("note"),
            "source": _UK["source"], "last_verified": _UK["last_verified"]}


def _us_tax(date_str, state):
    nat = {"ok": True, "country": "US", "tax_name": "Sales tax",
           "national_vat": 0.0, "date": date_str,
           "note": _US["note"], "source": _US["source"],
           "last_verified": _US["last_verified"]}
    if not state:
        nat["rate"] = 0.0
        nat["scope"] = "national (US has no VAT; supply a state for sales tax)"
        return nat
    state = state.upper()
    rates = _US["state_base_rates"]
    if state not in rates:
        return {"ok": False, "reason": f"unknown US state code: {state}"}
    nat["rate"] = rates[state]
    nat["scope"] = "state base rate (local taxes add on top)"
    nat["state"] = state
    if state in _US.get("rate_notes", {}):
        nat["state_note"] = _US["rate_notes"][state]
    return nat


def vat_rate(country: str = "GB", date: str = None, state: str = None) -> dict:
    """Look up the applicable consumption-tax rate.

    GB -> standard VAT for the given date (handles historical/temporary changes).
    US -> national VAT is 0; pass `state` (e.g. 'CA') for the state base sales-tax rate.
    """
    country = (country or "GB").upper()
    date = date or _dt.date.today().isoformat()
    if country == "GB":
        return _uk_vat(date)
    if country == "US":
        return _us_tax(date, state)
    return {"ok": False, "reason": f"country {country} not supported in v1 (GB, US only)"}

File: packages/localecheck-ts/src/vat.ts
// Tax-rate lookups: UK VAT by date, US sales tax by state.
// Reuses the same curated JSON datasets as the Python implementation.
import { readFileSync } from "node:fs";

const dataDir = new URL("../data/", import.meta.url);

interface UkBand { from: string; rate: number; note?: string }
interface UkData { standard_rate_history: UkBand[]; source: string; last_verified: string }
interface UsData {
  state_base_rates: Record<string, number>;
  rate_notes: Record<string, string>;
  note: string;
  source: string;
  last_verified: string;
}

const UK: UkData = JSON.parse(readFileSync(new URL("uk_vat.json", dataDir), "utf8"));
const US: UsData = JSON.parse(readFileSync(new URL("us_sales_tax.json", dataDir), "utf8"));

export interface TaxResult {
  ok: boolean;
  country?: string;
  tax_name?: string;
  rate?: number;
  date?: string;
  effective_from?: string;
  kind?: string;
  note?: string | null;
  national_vat?: number;
  scope?: string;
  state?: string;
  state_note?: string;
  source?: string;
  last_verified?: string;
  reason?: string;
}

function ukVat(date: string): TaxResult {
  if (!/^\d{4}-\d{2}-\d{2}$/.test(date)) return { ok: false, reason: `bad date: ${date}` };
  let applicable: UkBand | null = null;
  for (const band of UK.standard_rate_history) {
    if (band.from <= date) applicable = band;
    else break;
  }
  if (!applicable) {
    return { ok: false, reason: `no UK VAT rate recorded for ${date}` };
  }
  return {
    ok: true,
    country: "GB",
    tax_name: "VAT",
    rate: applicable.rate,
    date,
    effective_from: applicable.from,
    kind: "standard",
    note: applicable.note ?? null,
    source: UK.source,
    last_verified: UK.last_verified,
  };
}

function usTax(date: string, state?: string): TaxResult {
  const base: TaxResult = {
    ok: true,
    country: "US",
    tax_name: "Sales tax",
    national_vat: 0.0,
    date,
    note: US.note,
    source: US.source,
    last_verified: US.last_verified,
  };
  if (!state) {
    return { ...base, rate: 0.0, scope: "national (US has no VAT; supply a state for sales tax)" };
  }
  const st = state.toUpperCase();
  const rate = US.state_base_rates[st];
  if (rate === undefined) return { ok: false, reason: `unknown US state code: ${st}` };
  return {
    ...base,
    rate,
    scope: "state base rate (local taxes add on top)",
    state: st,
    state_note: US.rate_notes[st],
  };
}

export function vatRate(country = "GB", date?: string, state?: string): TaxResult {
  country = (country || "GB").toUpperCase();
  const d = date || new Date().toISOString().slice(0, 10);
  if (country === "GB") return ukVat(d);
  if (country === "US") return usTax(d, state);
  return { ok: false, reason: `country ${country} not supported in v1 (GB, US only)` };
}
